crates/hir_ty/src/diagnostics/expr.rs

```diff
@@ -317,6 +324,31 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
             }
         }
     }
+
+    fn validate_missing_tail_expr(
+        &mut self,
+        body_id: ExprId,
+        possible_tail_id: ExprId,
+        db: &dyn HirDatabase,
+    ) {
+        let mismatch = match self.infer.type_mismatch_for_expr(body_id) {
+            Some(m) => m,
+            None => return,
+        };
+
+        if let Some(possible_tail_ty) = self.infer.type_of_expr.get(possible_tail_id) {
+            if mismatch.actual == Ty::unit() && mismatch.expected == *possible_tail_ty {
+                let (_, source_map) = db.body_with_source_map(self.owner.into());
+
+                if let Ok(source_ptr) = source_map.expr_syntax(possible_tail_id) {
+                    self.sink.push(RemoveThisSemicolon {
+                        file: source_ptr.file_id,
+                        expr: source_ptr.value,
+                    });
+                }
+            }
+        }
+    }
```

Comment:
```suggestion
        let possible_tail_ty = if let Some(possible_tail_ty) = self.infer.type_of_expr.get(possible_tail_id) {
            possible_tail_ty
        } else {
            return;
        };

        if mismatch.actual != Ty::unit() || mismatch.expected != *possible_tail_ty {
            return;
        }

        let (_, source_map) = db.body_with_source_map(self.owner.into());
        if let Ok(source_ptr) = source_map.expr_syntax(possible_tail_id) {
            self.sink.push(RemoveThisSemicolon {
                file: source_ptr.file_id,
                expr: source_ptr.value,
            });
        }
    }
```

This significantly reduces the indentation.